Bluetooth: Tester: update tester to support BR GAP PTS test cases

lylezhu2012 opened the pull request:
The changes of the PR are used to support the BR GAP test cases.

lylezhu2012 committed:
Bluetooth: tester: Set setting `BREDR` and `SSP` if classic is enabled

Set the supported settings `BTP_GAP_SETTINGS_BREDR` and
`BTP_GAP_SETTINGS_SSP` if the Bluetooth classic is enabled.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support general and limited discoverable for classic

Disable discoverable mode if `cp->discoverable` is
`BTP_GAP_NON_DISCOVERABLE`.

Set general discoverable mode if `cp->discoverable` is
`BTP_GAP_GENERAL_DISCOVERABLE`.

Set limited discoverable mode if `cp->discoverable` is
`BTP_GAP_LIMITED_DISCOVERABLE`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support connectable mode for classic

Disable connectable mode for classic if `cp->connectable` is false.

Enable connectable mode for classic if `cp->connectable` is true.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add BR address type BTP_BR_ADDRESS_TYPE

The address type `BTP_BR_ADDRESS_TYPE=0xe0` of LE address is used to
mark as BR address.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support BR for security related commands

Look up the BR connection if the address type is BTP_BR_ADDRESS_TYPE.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: GAP: Support BR connection

According to the connection type of connection, get the appropriate
peer address.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support BR L2CAP listen mode

If cp->transport is BTP_L2CAP_TRANSPORT_BREDR, register BR L2CAP
server.

If cp->transport is not one of BTP_L2CAP_TRANSPORT_BREDR and
BTP_L2CAP_TRANSPORT_LE, return error code BTP_STATUS_FAILED.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support pincode request for BR

Set the pin code request callback `auth_pincode_entry`.

If `highsec` is false, call `bt_conn_auth_pincode_entry` with pin code
"0000".
If `highsec` is true, call `bt_conn_auth_pincode_entry` with pin code
"0000000000000000".

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Enable GAP discover feature for BR

Call `bt_br_discovery_cb_register` to register discovery callbacks.

Define a buffer br_adv_buf to report br discovery result.

Report device found event if discovery received callback triggered.

Start BR device discovery procedure if command `start_discovery` is
received.

Stop BR device discovery procedure if command `stop_discovery` is
received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support BR connect and disconnect features

Create BR connect if the address type is `BTP_BR_ADDRESS_TYPE` when
handling BTP GAP `connect` command.

Disconnect BR connect if the address type is `BTP_BR_ADDRESS_TYPE`
when handling BTP GAP `disconnect` command.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support BR L2CAP connect and disconnect BTP command

Change the channel id rage of BR L2CAP channel to
`ARRAY_SIZE(channels)` ~ `ARRAY_SIZE(channels) + CHANNELS`. The
unified channel ID can help identify whether the channel is BR or BLE
L2CAP channel.

When the address type of L2CAP connect command is
`BTP_BR_ADDRESS_TYPE`, create BR L2CAP channel connect request.

And if the channel ID falls into the range `ARRAY_SIZE(channels)`
~ `ARRAY_SIZE(channels) + CHANNELS` of L2CAP disconnect command,
create BR L2CAP disconnect request.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add command pair_v2 for BTP GAP

Compare with BTP GAP pair, the pair_v2 add more arguments, including
security mode, security level, and flags.

The argument `security mode` is used to set the security mode.

The argument `security level` is used to set the security level of the
specific security mode.

The argument `flags` is used to add additional setting, such as flag
`BTP_GAP_PAIR_V2_FLAG_FORCE_PAIR` is used to force the pairing
procedure.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Ignore error -EBUSY of `bt_conn_set_security`

The error code `-EBUSY` of `bt_conn_set_security` means the
pairing/encryption procedure is ongoing. Ignore the error code.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support l2cap connect rsp `insuff secure authen`

If the L2CAP connection response of command `listen` is
`BTP_L2CAP_CONNECTION_RESPONSE_INSUFF_SEC_AUTHEN`, set the security
level of L2CAP server to level 4.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add project configuration for BR

Set following configuration by default,
CONFIG_BT_CLASSIC=y
CONFIG_BT_CENTRAL=y
CONFIG_BT_PERIPHERAL=y
CONFIG_BT_SMP=y
CONFIG_BT_L2CAP_DYNAMIC_CHANNEL=y
CONFIG_BT_PAGE_TIMEOUT=0xFFFF

When building the project, the argument `--extra-conf prj_br.conf`
needs to be appended for BR.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2618,6 +2618,27 @@ struct bt_br_conn_param {
+ */
+struct bt_conn *bt_conn_lookup_addr_br(const bt_addr_t *peer);
+
+/** @brief Get destination (peer) address of a connection.
+ *
+ *  @param conn Connection object.
+ *
+ *  @return Destination address if @p conn is a valid @ref BT_CONN_TYPE_BR connection
+ */
+const bt_addr_t *bt_conn_get_dst_br(const struct bt_conn *conn);
```
You should rely on the `bt_conn_get_info` instead of introducing another function to retrieve this value

lylezhu2012 replied:
No. It is just similar with LE `bt_conn_get_dst_le`. I think `bt_conn_get_dst_br` is also required for BR.

For `bt_conn_get_info.`, if I want to get the peer address, I need to get a huge info. It is a burden for call stack. And for `bt_conn_get_info`, the key size will be always requested to read from controller, it will drag down the system if the peer is required in different function.

So, for just reading a peer address, it is not a good idea to call `bt_conn_get_info `.

Thalley replied:
@jhedberg any opinions here?

Thalley commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -481,6 +481,8 @@ int bt_le_create_conn_synced(const struct bt_conn *conn, const struct bt_le_ext_
 			     uint8_t subevent);
 
 bool bt_addr_le_is_bonded(uint8_t id, const bt_addr_le_t *addr);
+bool bt_addr_br_is_bonded(const bt_addr_t *addr);
```
Consider making this public similar to https://github.com/zephyrproject-rtos/zephyr/pull/85590

lylezhu2012 replied:
Yes. I also considered it. But I did not find your PR, before, when I was adding this function.

lylezhu2012 replied:
Created a PR for it https://github.com/zephyrproject-rtos/zephyr/pull/87048

Thalley reviewed (changes requested):
Not a complete review, but did check changes to common files

lylezhu2012 commented on tests/bluetooth/tester/src/btp/btp.h:
```diff
@@ -43,6 +43,8 @@
 #define BTP_MTU 1024
 #define BTP_DATA_MAX_SIZE (BTP_MTU - sizeof(struct btp_hdr))
 
+#define BTP_BR_ADDRESS_TYPE 0xe0
```
btp.h is a common file for BR and BLE. Because, current BTP services (such as GAP, L2CAP) are used by BR. And all services should use same service id range. It is better to put them to same place. Than from the perspective of LE, it should know that the type `0xe0` has been assigned to BR.

Thalley replied:
Yeah, but my point is that btp.h is too generic of a file to define something like an address type in. I think something like `btp_gap.h` (which is AFAIK also used by BR), is a better place, but I'll defer to @sjanc on this

lylezhu2012 replied:
Got. Let check whether it only be used by GAP.

lylezhu2012 replied:
Ohh, I found it also used by L2CAP. I am not sure whether it is a reasonable that it includes "btp_gap.h" in "btp_l2cap.c".

Thalley replied:
That's a good point :) Hopefully Szymon has an idea of where to put it then :D

sjanc replied:
for now lets keep it here,   I need to do a cleanup on BTP docs regarding address type anyway, so after that we will sync tester definitions with docs (although I think it will end up in btp.h anyway)

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -505,6 +569,15 @@ static uint8_t set_connectable(const void *cmd, uint16_t cmd_len,
 	const struct btp_gap_set_connectable_cmd *cp = cmd;
 	struct btp_gap_set_connectable_rp *rp = rsp;
 
+#if defined(CONFIG_BT_CLASSIC)
```
Yes. Let me update it.

lylezhu2012 replied:
I have updated all of them.

lylezhu2012 committed:
Bluetooth: tester: Support board mimxrt1170_evk@B/mimxrt1176/cm7

Add test case bluetooth.general.tester_br_edr.

Add overlay `zephyr,sram` and `zephyr,uart-pipe` for the board.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -38,7 +38,11 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
 #if defined(CONFIG_BT_EXT_ADV)
 #define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + 2 * CONFIG_BT_EXT_SCAN_BUF_SIZE)
 #else
+#if defined(CONFIG_BT_CLASSIC)
+#define ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + BT_BR_EIR_SIZE)
```
What happens in a dual-mode build here? 

If both `CONFIG_BT_EXT_ADV` and `CONFIG_BT_CLASSIC` is enabled, then this doesn't look like it will be properly set

lylezhu2012 replied:
The default value of `CONFIG_BT_EXT_SCAN_BUF_SIZE` is 229. So the buffer length is bigger than the length required by the BR. But the value `CONFIG_BT_EXT_SCAN_BUF_SIZE`  can be configured by application layer.

I will added a buffer dedicated to BR.

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -102,7 +106,14 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
 	struct bt_conn_info info;
 
-	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr_str, sizeof(addr_str));
+	memset(addr_str, 0, sizeof(addr_str));
```
```suggestion
	(void)memset(addr_str, 0, sizeof(addr_str));
```

jhedberg replied:
I really like the `__must_check` annotation, and I wish we used it more. When we take the policy of requiring explicit typecasts to ignore returns from non-void functions, we *lose* the ability to distinguish between APIs where it's always a bug to ignore the return value (i.e. where we'd use `__must_check`), and where it might be a bug but also might not be (in the particular case of `memset()` it's almost *never* a bug to ignore it). I know this is all coming from approved coding guidelines, but I believe it's actually making the project worse rather than better as far as code quality is concerned.

Thalley replied:
I tend to agree with you on this - I never liked the typecast to (void) as it just adds noise and it's hard to determine if the function may return an error (where casting to (void) is not allowed). 

Nevertheless, if Zephyr wants to be MISRA compliant, we have to follow that rule. This rule for memset and memcpy has been discussed in many forums over the years, and it's just not great. 

Work arounds would be have to have z_memset that does not have a `void *` return value, but that also just doesn't seem great. 

@simhein do you have opinion on this one?

jhedberg replied:
The main challenge here is that the definition APIs like `memset()` are external, and we can't change them. If we could, introducing a new API annotation called something like `__may_ignore` (basically the inverse of `__must_check`) would let us explicitly document the cases there it's perfectly normal to want to ignore the return value, and for such cases we wouldn't require an explicit `(void)` cast.

lylezhu2012 replied:
For Now, I added `(void)` to ignore the return value to avoid the MISRA issue. If there is a zephyr specific `memset()`, I think we can replace it globally.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -111,10 +122,19 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
-	ev.interval = sys_cpu_to_le16(info.le.interval);
-	ev.latency = sys_cpu_to_le16(info.le.latency);
-	ev.timeout = sys_cpu_to_le16(info.le.timeout);
+	memset(&ev, 0, sizeof(ev));
```
```suggestion
	(void)memset(&ev, 0, sizeof(ev));
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -111,10 +122,19 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
+			ev.latency = sys_cpu_to_le16(info.le.latency);
+			ev.timeout = sys_cpu_to_le16(info.le.timeout);
+		}
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC)) {
```
Missing check for `BT_CONN_TYPE_BR`

lylezhu2012 replied:
Fixed.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -130,13 +150,24 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
 
-	bt_addr_le_copy(&ev.address, addr);
+		addr = bt_conn_get_dst(conn);
+		(void)bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
```
Ditto; shall test return value

lylezhu2012 replied:
The return value of `bt_addr_le_to_str()` is ignored in the whole of the file. And it is only used for logging. 

To avoid possible issues, the peer address can only be printed after checking the return value of `bt_addr_le_to_str()`. If do this change, the purpose of log may not be achieved, because the log may not be printed sometimes due to the returned value of `bt_addr_le_to_str()` is not a positive value. 
Or I suggest removing all the codes that formats the address, instead adding a function to print the address directly.

Thalley replied:
> If do this change, the purpose of log may not be achieved, because the log may not be printed sometimes due to the returned value of bt_addr_le_to_str() is not a positive value.

If it returns an error, the there won't be a address string to log anyhow. So in that regard, it should still be checked. 

However, looking at `bt_addr_le_to_str` it not only does not document any possible errors, but it should also ensure that no encoding errors, so perhaps `bt_addr_le_to_str` should be made `void` really. @jhedberg thoughts? 

I won't block changes to the tester on MISRA rules, as it's not really core to Zephyr, and will leave it up to the maintainer, @sjanc, to decide whether to be MISRA compliant / strictly follow all coding guidelines in the BT Tester

sjanc replied:
it is for debug logging only so I'm fine with explicit void cast here  (we could consider asserting but than this should be for all tester code)

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -130,13 +150,24 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
+		(void)bt_addr_to_str(br_addr, addr_str, sizeof(addr_str));
```
Ditto

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -182,14 +217,35 @@ static bool le_param_req(struct bt_conn *conn, struct bt_le_conn_param *param)
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
```
`bt_conn_get_dst` may return NULL

lylezhu2012 replied:
It has been checked by calling `bt_conn_is_type()` before calling `bt_conn_get_dst()`. So the `NULL` should be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -182,14 +217,35 @@ static bool le_param_req(struct bt_conn *conn, struct bt_le_conn_param *param)
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
`bt_conn_get_dst_br` may return NULL

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -616,15 +683,58 @@ static uint8_t set_discoverable(const void *cmd, uint16_t cmd_len,
+
+			err = bt_br_set_connectable(true);
+			if (err == -EALREADY) {
+				(void)bt_br_set_discoverable(false, false);
```
Shall test return value

lylezhu2012 replied:
This is intentional. Its return value, success or failure, has no effect on subsequent execution.

Thalley replied:
It is still a coding guideline violation :) As mentioned, will defer the strict enforcement of those guidelines to the maintainer (@sjanc ) here

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -616,15 +683,58 @@ static uint8_t set_discoverable(const void *cmd, uint16_t cmd_len,
+
+			err = bt_br_set_connectable(true);
+			if (err == -EALREADY) {
+				(void)bt_br_set_discoverable(false, false);
```
Shall test return value

lylezhu2012 replied:
This is intentional. Its return value, success or failure, has no effect on subsequent execution.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1021,17 +1131,90 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t evtype,
+	ev->eir_data_len = 0;
+	ev->rssi = result->rssi;
+	ev->flags = BTP_GAP_DEVICE_FOUND_FLAG_AD | BTP_GAP_DEVICE_FOUND_FLAG_RSSI;
+	for (uint32_t i = 0U; i < sizeof(result->eir); i = ev->eir_data_len) {
```
```suggestion
	for (size_t i = 0U; i < sizeof(result->eir); i = ev->eir_data_len) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1021,17 +1131,90 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t evtype,
+			break;
+		}
+	}
+	memcpy(net_buf_simple_add(adv_buf, ev->eir_data_len), result->eir, ev->eir_data_len);
```
```suggestion
	(void)memcpy(net_buf_simple_add(adv_buf, ev->eir_data_len), result->eir, ev->eir_data_len);
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1021,17 +1131,90 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t evtype,
+
+	err = bt_br_discovery_start(&br_discover_param, br_discover_result,
+				    ARRAY_SIZE(br_discover_result));
+	if (err) {
```
```suggestion
	if (err != 0) {
```

Expression in if statements shall be boolean as per rule 85 of https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

lylezhu2012 replied:
I didn't realize that zephyr started having MISRA requirements for the code.

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1059,6 +1250,28 @@ static uint8_t stop_discovery(const void *cmd, uint16_t cmd_len,
+	struct bt_conn *conn;
+
+	conn = bt_conn_create_br(&cp->address.a, BT_BR_CONN_PARAM_DEFAULT);
+	if (!conn) {
```
Expressions in `if` statements shall be boolean as per rule 85 of https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

```suggestion
	if (conn == NULL) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1122,9 +1348,20 @@ static uint8_t disconnect(const void *cmd, uint16_t cmd_len,
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
+		bt_addr_le_copy(&ev.address, addr);
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
`get_dst` may return NULL

lylezhu2012 replied:
In current situation, `NULL` should not be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1133,19 +1370,40 @@ static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
+		bt_addr_le_copy(&ev.address, addr);
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
`get_dst` may return NULL

lylezhu2012 replied:
In this situation, the transport is `BT_CONN_TYPE_BR`, the `NULL` should not be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1133,19 +1370,40 @@ static void auth_passkey_display(struct bt_conn *conn, unsigned int passkey)
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
+		bt_addr_le_copy(&ev.address, addr);
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
get_dst may return NULL

lylezhu2012 replied:
In this situation, the transport is `BT_CONN_TYPE_BR`, the `NULL` should not be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1160,19 +1418,33 @@ enum bt_security_err auth_pairing_accept(struct bt_conn *conn,
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
+		if (!bt_le_bond_exists(BT_ID_DEFAULT, addr)) {
+			return BT_SECURITY_ERR_SUCCESS;
+		}
+
+		bt_addr_le_copy(&ev.address, addr);
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
get_dst may return NULL

lylezhu2012 replied:
In this situation, the transport is `BT_CONN_TYPE_BR`, the `NULL` should not be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1181,9 +1453,20 @@ enum bt_security_err auth_pairing_accept(struct bt_conn *conn,
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		const bt_addr_le_t *addr;
+
+		addr = bt_conn_get_dst(conn);
+		bt_addr_le_copy(&ev.address, addr);
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		const bt_addr_t *br_addr;
+
+		br_addr = bt_conn_get_dst_br(conn);
```
get_dst may return NULL

lylezhu2012 replied:
In this situation, the transport is `BT_CONN_TYPE_BR`, the `NULL` should not be returned.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1257,13 +1567,70 @@ static uint8_t pair(const void *cmd, uint16_t cmd_len,
+		conn = bt_conn_lookup_addr_le(BT_ID_DEFAULT, &cp->address);
+	}
+
+	if (!conn) {
```
```suggestion
	if (conn == NULL) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+
+	ev.chan_id = chan->chan_id;
+	/* TODO: ev.psm */
+	if (!bt_conn_get_info(l2cap_chan->conn, &info)) {
```
```suggestion
	if (bt_conn_get_info(l2cap_chan->conn, &info) == 0) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+	/* TODO: ev.result */
+	ev.chan_id = chan->chan_id;
+	/* TODO: ev.psm */
+	if (!bt_conn_get_info(l2cap_chan->conn, &info)) {
```
```suggestion
	if (bt_conn_get_info(l2cap_chan->conn, &info) == 0) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+	uint8_t i;
+	struct br_channel *chan;
+
+	for (i = 0U; i < CHANNELS; i++) {
```
Can `CHANNELS` become > 0xFF?

lylezhu2012 replied:
In current implementation, the type of channel id is `uint8_t`. And the type of BTP channel id is also `uint8_t`.

But for dual mode, the channel id should not be more than `0x7f`. I will add comments for the macro of `CHANNELS` to avoid the `CHANNELS` is more than `0x7f`.

Thalley replied:
> Added comments to macro `CHANNELS`.

You can also consider adding a `BUILD_ASSERT` to make it extremely clear that it shall not be above 0x7F, but a comment is also fine :)

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+	int err;
+
+	conn = bt_conn_lookup_addr_br(&cp->address.a);
+	if (!conn) {
```
```suggestion
	if (conn == NULL) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+	}
+
+	br_chan = get_free_br_channel();
+	if (!br_chan) {
```
```suggestion
	if (br_chan == NULL) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +241,140 @@ static struct channel *get_free_channel()
+	return BTP_STATUS_SUCCESS;
+
+fail:
+	if (br_chan) {
```
```suggestion
	if (br_chan != NULL) {
```

lylezhu2012 replied:
udpated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -308,13 +454,49 @@ static uint8_t connect(const void *cmd, uint16_t cmd_len,
+	br_chan = &br_channels[chan_id];
+
+	err = bt_l2cap_chan_disconnect(&br_chan->br.chan);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

lylezhu2012 replied:
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -513,15 +695,47 @@ static int accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+	}
+
+	chan = get_free_br_channel();
+	if (!chan) {
```
```suggestion
	if (chan == NULL) {
```

lylezhu2012 replied:
Updated.

lylezhu2012 committed:
Bluetooth: tester: Add BTP event encryption change

Report encryption change event when callback `le_security_changed` is
triggered.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -51,6 +51,9 @@ struct bt_br_discovery_priv {
 };
 
+/* Maximum size of Extended Inquiry Response */
+#define BT_BR_EIR_SIZE 240
```
Updated. Change `BT_BR_EIR_SIZE` to `BT_BR_EIR_SIZE_MAX`

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -51,6 +51,9 @@ struct bt_br_discovery_priv {
 	bool resolving;
 };
 
+/* Maximum size of Extended Inquiry Response */
```
Updated.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -102,7 +106,14 @@ static void le_connected(struct bt_conn *conn, uint8_t err)
+	memset(addr_str, 0, sizeof(addr_str));
+
+	if (bt_conn_is_type(conn, BT_CONN_TYPE_LE)) {
+		(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr_str, sizeof(addr_str));
+	} else if (IS_ENABLED(CONFIG_BT_CLASSIC) && bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		(void)bt_addr_to_str(bt_conn_get_dst_br(conn), addr_str, sizeof(addr_str));
```
Updated.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2904,6 +2904,20 @@ const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn)
 	return &conn->le.dst;
 }
 
+const bt_addr_t *bt_conn_get_dst_br(const struct bt_conn *conn)
+{
+#if defined(CONFIG_BT_CLASSIC)
+	if (!bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
+		LOG_DBG("Invalid connection type: %u for %p", conn->type, conn);
+		return NULL;
+	}
+
+	return &conn->br.dst;
```
Consider adding a `if (conn == NULL)` check

lylezhu2012 replied:
I found it has been checked in `bt_conn_is_type()`. So I think it is not necessary.

Thalley replied:
If `conn == NULL` and `bt_conn_is_type` fails, then you dereference `conn` in the log statement `LOG_DBG("Invalid connection type: %u for %p", conn->type, conn);`

lylezhu2012 replied:
Let me update it.

Thalley reviewed (commented):
Changes are OK with me, but would like input from @sjanc on the coding guidelines compliance and location of `BTP_BR_ADDRESS_TYPE`

lylezhu2012 commented:
Rebase the PR to fix the conflict.

@sjanc , Have you reviewed it? Any comments? You know this PR has been created for almost a month. Now it is blocked and waiting for your review. All of our future PTS support for Classic modules will be based on the results of this PR. Only after this PR is completed can we start enabling PTS for other Classic modules. Now that this PR is not completed, all our work is blocked.

lylezhu2012 commented:
Updated the PR according to the comments at https://github.com/auto-pts/auto-pts/pull/1407.

lylezhu2012 commented:
> Updated the PR according to the comments at [auto-pts/auto-pts#1407](https://github.com/auto-pts/auto-pts/pull/1407).

The changes of [auto-pts/auto-pts#1407](https://github.com/auto-pts/auto-pts/pull/1407) has been merged into main branch of auto-pts by  [auto-pts/auto-pts#1435]https://github.com/auto-pts/auto-pts/pull/1435.

sjanc commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -41,6 +41,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
 #endif
 
+#if defined(CONFIG_BT_CLASSIC)
+#define BR_ADV_BUF_LEN (sizeof(struct btp_gap_device_found_ev) + BT_BR_EIR_SIZE_MAX)
```
nitpick: I'd name this BR_EIR_BUF_LEN  (there is no advertising data in br/edr)

sjanc commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -579,15 +652,64 @@ static uint8_t set_discoverable(const void *cmd, uint16_t cmd_len,
+			if (err < 0) {
+				return BTP_STATUS_FAILED;
+			}
+			ad_flags &= ~BT_LE_AD_NO_BREDR;
```
I assume that there is no option to disable BR/EDR on runtime?  (ie BTP_GAP_SETTINGS_BREDR is always set when BR/EDR is compiled in?)

sjanc commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1165,6 +1480,24 @@ static struct bt_conn_auth_info_cb auth_info_cb = {
+static void auth_pincode_entry(struct bt_conn *conn, bool highsec)
+{
+	const char *pin = "0000";
+	const char *pin16 = "0000000000000000";
```
this suggests that there might be something missing on BTP side (for now I'm fine with fixed pins since I suppose this can be set via IXIT on PTS side)

should there be event + cmd for this? similar to passkey entry?

sjanc commented on tests/bluetooth/tester/src/btp_l2cap.c:
```diff
@@ -233,6 +242,146 @@ static struct channel *get_free_channel()
+	struct br_channel *chan = CONTAINER_OF(l2cap_br_chan, struct br_channel, br);
+	struct bt_conn_info info;
+
+	(void)memset(&ev, 0, sizeof(struct btp_l2cap_disconnected_ev));
```
nitpick: I'd suggest to stick to  memset(&ev, 0, sizeof(ev)) pattern

sjanc reviewed (commented):
overall looks ok to me

Thalley reviewed (approved):
No remaining blockers from my side. Only reviewed common files

sjanc reviewed (approved):
apparently I didn't approve it yet :)

any nitpicks may be fixed in follow-ups
